From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: ProstoyArtemka <ProstoyArtemka@yandex.ru>
Date: Thu, 25 Jan 2024 21:06:11 +0700
Subject: [PATCH] Yes yes yes


diff --git a/src/main/java/net/minecraft/server/level/ServerPlayerGameMode.java b/src/main/java/net/minecraft/server/level/ServerPlayerGameMode.java
index 58f972832c39a27a8ccd606f9144e1c54adbf6f3..b8166b661de82bd1b7195ed884433967562acde4 100644
--- a/src/main/java/net/minecraft/server/level/ServerPlayerGameMode.java
+++ b/src/main/java/net/minecraft/server/level/ServerPlayerGameMode.java
@@ -27,6 +27,7 @@ import net.minecraft.world.level.block.state.BlockState;
 import net.minecraft.world.level.block.state.properties.DoubleBlockHalf;
 import net.minecraft.world.phys.BlockHitResult;
 import net.minecraft.world.phys.Vec3;
+import org.ospaindustries.block.CustomBlock;
 import org.slf4j.Logger;
 
 // CraftBukkit start
@@ -374,6 +375,12 @@ public class ServerPlayerGameMode {
 
             this.level.getCraftServer().getPluginManager().callEvent(event);
 
+            if (bblock.isCustom() && !event.isCancelled()) {
+                CustomBlock customBlock = bblock.toCustom();
+
+                customBlock.breakBlock(true, this.player.getBukkitEntity(), true);
+            }
+
             if (event.isCancelled()) {
                 if (isSwordNoBreak) {
                     return false;
diff --git a/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java b/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
index 2d40fc5b9d1e6461c431e3703236ba0d26eb6560..de28e5006c38e3b43aa5304731f85cf0ec0fa366 100644
--- a/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
+++ b/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
@@ -241,6 +241,7 @@ import org.bukkit.inventory.InventoryView;
 import org.bukkit.inventory.Recipe;
 import org.bukkit.inventory.meta.BookMeta;
 import org.bukkit.potion.PotionEffect;
+import org.bukkit.scheduler.BukkitRunnable;
 import org.bukkit.util.Vector;
 import org.ospaindustries.block.CustomBlock;
 import org.ospaindustries.entity.CustomEntity;
@@ -451,18 +452,12 @@ public class CraftEventFactory {
         BlockPlaceEvent event = new BlockPlaceEvent(placedBlock, replacedBlockState, blockClicked, item, player, canBuild, equipmentSlot);
         craftServer.getPluginManager().callEvent(event);
 
-        if (item.isCustom()) {
-            CustomItem customItem = item.toCustomItem();
-            if (customItem == null) return event;
+        if (item.isCustom() && item.toCustomItem().isBlockItem()) {
+            Location placedLocation = placedBlock.getLocation();
+            CustomBlock customBlock = CustomBlock.fromBlockItem(item.toCustomItem());
 
-            if (customItem.isBlockItem()) {
-                CustomBlock customBlock = CustomBlock.fromBlockItem(customItem);
-
-                if (customBlock == null) return event;
-
-                customBlock.place(placedBlock.getLocation());
-                customBlock.getBehavior().placeBlock(event);
-            }
+            customBlock.place(placedLocation);
+            customBlock.getBehavior().placeBlock(event);
         }
 
         return event;
@@ -567,6 +562,9 @@ public class CraftEventFactory {
     public static PlayerInteractEvent callPlayerInteractEvent(net.minecraft.world.entity.player.Player who, Action action, BlockPos position, Direction direction, ItemStack itemstack, boolean cancelledBlock, InteractionHand hand, Vec3 targetPos) {
         return CraftEventFactory.callPlayerInteractEvent(who, action, position, direction, itemstack, cancelledBlock, false, hand, targetPos);
     }
+
+    private static ArrayList<Player> players = new ArrayList<>();
+
     public static PlayerInteractEvent callPlayerInteractEvent(net.minecraft.world.entity.player.Player who, Action action, BlockPos position, Direction direction, ItemStack itemstack, boolean cancelledBlock, boolean cancelledItem, InteractionHand hand, Vec3 targetPos) {
         // Paper end - cancelledItem param
         Player player = (who == null) ? null : (Player) who.getBukkitEntity();
@@ -714,10 +712,10 @@ public class CraftEventFactory {
 
         if (blockClicked.isCustom()) {
             CustomBlock customBlock = blockClicked.toCustom();
-            if (!customBlock.exist()) return null;
+            if (!customBlock.exist()) return event;
 
-            BreakBlockRunnable.breakBlockRunnables.get(blockClicked.getLocation()).cancel();
-            BreakBlockRunnable.breakBlockRunnables.remove(blockClicked.getLocation());
+            BreakBlockRunnable.breakBlockRunnables.get(Location.locationToString(blockClicked.getLocation().toBlockLocation())).cancel();
+            BreakBlockRunnable.breakBlockRunnables.remove(Location.locationToString(blockClicked.getLocation().toBlockLocation()));
 
             customBlock.getBehavior().stopBreaking(event);
         }
